Move the metrics collection documents (#3533)
SumoLogic/sumologic-documentation: docs/send-data/kubernetes/collecting-application-metrics.md

---
id: collecting-application-metrics
title: Collecting Application Metrics
sidebar_label: Collecting Application Metrics
description: This document covers multiple different use cases related to scraping custom application metrics exposed in Prometheus format.
---

This document covers multiple different use cases related to scraping custom application metrics exposed in Prometheus format.

There are two major sections:

> - [Scraping metrics](#scraping-metrics) which describes how to send your application metrics to sumo

Reviewer: ```suggestion
- [Scraping metrics](#scraping-metrics) which describes how to send your application metrics to Sumo Logic
```

- [Metrics modifications](#metrics-modifications) which describes how to filter metrics and rename both metrics and metric metadata

## Scraping metrics

This section describes how to scrape metrics from your applications. The following scenarios are covered:

- [Application metrics are exposed (one endpoint scenario)](#application-metrics-are-exposed-one-endpoint-scenario)

> - [Application metrics are exposed (multiple endpoints scenario)](#application-metrics-are-exposed-multiple-enpoints-scenario)

Reviewer: ```suggestion
- [Application metrics are exposed (multiple endpoints scenario)](#application-metrics-are-exposed-multiple-endpoints-scenario)
```

- [Application metrics are not exposed](#application-metrics-are-not-exposed)

### Application metrics are exposed (one endpoint scenario)

If there is only one endpoint in the Pod you want to scrape metrics from, you can use annotations. Add the following annotations to your Pod
definition:

```yaml
# ...
annotations:
  prometheus.io/port: 8000 # Port which metrics should be scraped from
  prometheus.io/scrape: true # Set if metrics should be scraped from this Pod
  prometheus.io/path: "/metrics" # Path which metrics should be scraped from
```

> **NOTE:** If you add more than one annotation with the same name, only the last one will be used.

Reviewer: ```suggestion
:::note
If you add more than one annotation with the same name, only the last one will be used.
:::
```

```yaml
sumologic:
  metrics:
    additionalServiceMonitors:
      - name: my-custom-app-service-monitor
        endpoints:
          - port: some-port
            path: /metrics
          - port: another-port
            path: /custom-endpoint
        namespaceSelector:
          matchNames:
            - my-custom-app-namespace
        selector:
          matchLabels:
            app: my-custom-app-service
```

### Application metrics are not exposed

In case you want to scrape metrics from an application which does not expose a Prometheus endpoint, you can use telegraf operator. It will
scrape metrics according to configuration and expose them on port `9273` so Prometheus will be able to scrape them.

For example to expose metrics from nginx Pod, you can use the following annotations:

```yaml
annotations:
  telegraf.influxdata.com/inputs: |+
  [[inputs.nginx]]
  urls = ["http://localhost/nginx_status"]
  telegraf.influxdata.com/class: sumologic-prometheus
  telegraf.influxdata.com/limits-cpu: '750m'
```

`sumologic-prometheus` defines the way telegraf operator will expose the metrics. They are going to be exposed in prometheus format on port
`9273` and `/metrics` path.

> **NOTE** If you apply annotations on Pod which is owned by another object, e.g. DaemonSet, it won't take affect. In such case, the
> annotation should be added to Pod specification in DaemonSet template.

Reviewer: ```suggestion
:::note
If you apply annotations on Pod which is owned by another object, for example DaemonSet, it won't take affect. In such case, the annotation should be added to Pod specification in DaemonSet template.
:::
```

After restart, the Pod should have an additional `telegraf` container.

To scrape and forward exposed metrics to Sumo Logic, please follow one of the following scenarios:

- [Application metrics are exposed (one endpoint scenario)](#application-metrics-are-exposed-one-endpoint-scenario)

> - [Application metrics are exposed (multiple endpoints scenario)](#application-metrics-are-exposed-multiple-enpoints-scenario)

Reviewer: ```suggestion
- [Application metrics are exposed (multiple endpoints scenario)](#application-metrics-are-exposed-multiple-endpoints-scenario)
```

| k8s.pod.name              | Kubernetes Pod name                                        |
| k8s.pod.pod_name          | Kubernetes Pod name                                        |
| k8s.replicaset.name       | Kubernetes Replicaset name                                 |
| k8s.service.name          | Kubernetes Service name                                    |
| k8s.statefulset.name      | Kubernetes Statefulset name                                |
| pod*labels*\<label_name\> | Kubernetes Pod label. Every label is a different attribute |
| prometheus                | Prometheus                                                 |
| prometheus_replica        | Prometheus Replica name                                    |
| prometheus_service        | Prometheus Service name                                    |

:::note
Before ingestion to Sumo Logic, attributes are renamed according to the [sumologicschemaprocessor documentation][sumologicschema]
:::

[sumologicschema]:
  https://github.com/SumoLogic/sumologic-otel-collector/tree/main/pkg/processor/sumologicschemaprocessor#attribute-translation

### Renaming metric

> In order to rename metrics, the [transformprocessor] can be use. Please look at the following snippet:

Reviewer: ```suggestion
In order to rename metrics, the [transformprocessor] can be used. Please look at the following snippet:
```

```yaml
sumologic:
  metrics:
    otelcol:
      extraProcessors:
        - transform/1:
            metric_statements:
              - context: metric
                statements:
                  ## Renames <old_name> to <new_name>
                  - set(name, "<new_name>") where name == "<old_name>"
```

### Adding or renaming metadata

> If you want to add or rename metadata, the [transformprocessor] can be use. Please look at the following snippet:

Reviewer: ```suggestion
If you want to add or rename metadata, the [transformprocessor] can be used. Please look at the following snippet:
```

```yaml
sumologic:
  metrics:
    otelcol:
      extraProcessors:
        - transform/1:
            metric_statements:
              - context: resource
                statements:
                  ## adds <new_name> metadata
                  - set(attributes["<new_name>"], attributes["<old_name>"])
                  ## adds <new_static_name> metadata
                  - set(attributes["<new_static_name>"], "<static_value>")
                  ## removes <old_name> metadata
                  - delete_key(attributes, "<old_name>")
```

**Note:** See [Default attributes](#default-attributes) for more information about attributes.

[transformprocessor]: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/transformprocessor

## Investigation

If you do not see your metrics in Sumo Logic, please check the following stages:

- [Collecting Application Metrics](collecting-application-metrics.md)
  - [Scraping metrics](#scraping-metrics)
    - [Application metrics are exposed (one endpoint scenario)](#application-metrics-are-exposed-one-endpoint-scenario)

>     - [Application metrics are exposed (multiple endpoints scenario)](#application-metrics-are-exposed-multiple-enpoints-scenario)

Reviewer: ```suggestion
    - [Application metrics are exposed (multiple endpoints scenario)](#application-metrics-are-exposed-multiple-endpoints-scenario)
```

      - [Example](#example)
    - [Application metrics are not exposed](#application-metrics-are-not-exposed)
  - [Metrics modifications](#metrics-modifications)
    - [Filtering metrics](#filtering-metrics)
      - [Default attributes](#default-attributes)
    - [Renaming metric](#renaming-metric)
    - [Adding or renaming metadata](#adding-or-renaming-metadata)
  - [Investigation](#investigation)
    - [Check if metrics are in Prometheus](#check-if-metrics-are-in-prometheus)
      - [Investigate Prometheus scrape configuration](#investigate-prometheus-scrape-configuration)
        - [Pod is visible in Prometheus targets](#pod-is-visible-in-prometheus-targets)
        - [There is no target for serviceMonitor](#there-is-no-target-for-servicemonitor)

>         - [Pod is not visible in target for custom serviceMonitor](#pod-is-not-visible-in-target-for-custom-servicemonitor)

Reviewer 2: I don't think we need this whole TOC here. Telling the user to make sure they followed the steps from the start of this document should be enough.

Reviewer: Agreed.

> ### Check if metrics are in Prometheus

Reviewer 2: This is all specific to Prometheus and shouldn't be here for v4. If we still want an investigation section here, it should use the otel specific troubleshooting steps @sumo-drosiek recently added to the Chart documentation.

Author: I removed this part. I believe that troubleshooting part should belong to troubleshooting document si I will leave it there.
